File: c++/problem_004.cpp
#include<time.h>
#include<stdio.h>
#include<string.h>
#include<string>
#include<vector>
#include<algorithm>

void blurb()
{
    const char * blurb = R"(
    A palindromic number reads the same both ways. The largest palindrome made
    from the product of two 2-digit numbers is 9009 = 91 × 99.

    Find the largest palindrome made from the product of two 3-digit numbers.
    )";
    printf("%s \n", blurb);
}

std::vector<int> find_palindromic_numbers(int lower_bound, int upper_bound)
{
    std::vector<int> palindromic_numbers;
    for (int a = upper_bound; a >= lower_bound ; a = a - 1) {
        std::string aString = std::to_string(a);
        std::string reversedAString = aString;
        std::reverse(reversedAString.begin(), reversedAString.end());
        if (aString == reversedAString) {
            palindromic_numbers.push_back(a);
        }
    }
    return palindromic_numbers;
}

int find_largest_factor(std::vector<int> nos, int lower_bound, int upper_bound)
{
    for (std::vector<int>::iterator it = nos.begin(); it != nos.end() ; ++it) {
        int no = *it;
        for (int j = upper_bound; j >= lower_bound; j--) {
            if (no % j == 0) {
                int divisor = no / j;
                if (lower_bound <= divisor && divisor <= upper_bound) {
                    printf("%d * %d == %d\n", j, divisor, no);
                    return no;
                }
            }
        }
    }
    return 0;
}

int problem_004(int lower_bound, int upper_bound)
{
    std::vector<int> palindromic_numbers = find_palindromic_numbers(
                                               lower_bound * lower_bound,
                                               upper_bound * upper_bound);
    return find_largest_factor(palindromic_numbers, lower_bound, upper_bound);
}

int main()
{
    blurb();
    clock_t begin = clock();
    int result = problem_004(100, 999);
    clock_t end = clock();
    float elapsed_secs = ((float)end - (float)begin) / 1000000.0F;
    printf("Result: %d \n", result);
    printf("Completed in: %f seconds. \n", elapsed_secs);
}

File: javascript/problem_004.js
#!/usr/bin/env node

function printBlurb() {
    var blurb = `
    A palindromic number reads the same both ways. The largest palindrome made
    from the product of two 2-digit numbers is 9009 = 91 × 99.

    Find the largest palindrome made from the product of two 3-digit numbers.
    `
    console.log(blurb)
}

function reverseString(string) {
    var reversedString = string.split("").reverse().join("")
    return reversedString
}

function findPalindromicNumbers(lowerBound, upperBound) {
    var palindromicNumbers = []
    for (var i = lowerBound ; i < upperBound ; i ++) {
        iString = i.toString()
        if (iString == reverseString(iString)) {
            palindromicNumbers.push(i)
        }
    }
    return palindromicNumbers

}

function numSort(a, b) {
    return a - b
}

function findLargestFactor(palindromicNumbers, lowerBound, upperBound) {
    var palindromicNumbersSorted = palindromicNumbers.sort(numSort).reverse()
    for (var i = 0 ; i < palindromicNumbersSorted.length ; i++) {
        var palindromicNumber = palindromicNumbersSorted[i]
        for (var j = upperBound ; j >= lowerBound ; j--) {
            if (palindromicNumber % j == 0) {
                var divisor = palindromicNumber / j
                if (lowerBound <= divisor && divisor <= upperBound) {
                    console.log("%d * %d == %d\n", j, divisor, palindromicNumber)
                    return palindromicNumber
                }
            }
        }
    }
}

function problem_004(lowerBound, upperBound) {
    palindromicNumbers = findPalindromicNumbers(lowerBound*lowerBound, upperBound*upperBound)
    return findLargestFactor(palindromicNumbers, lowerBound, upperBound)
}

printBlurb()
console.time('Completed in')
result = problem_004(100, 999)
console.log(`Result: ${result}`)
console.timeEnd('Completed in')
